Optimizations

gabi-250 opened the pull request:
This branch contains an optimized version of the interpreter. The improvements include the removal of the constants dict and of the calls to `get_block` in the interpreter.

gabi-250 committed:
Add get_printable_location.

gabi-250 committed:
Add _immutable_fields_.

gabi-250 committed:
Add immutable fields, jit hints.

ltratt commented on llvm_objects.py:
```diff
@@ -108,6 +113,38 @@ def _set_block(self, l_block, w_block):
+
+                    instruction.w_bb_true = self.get_block(l_block_true)
+                    instruction.w_bb_false = self.get_block(l_block_false)
+
```
Stray blank line.

ltratt commented on llvm_objects.py:
```diff
@@ -117,7 +154,11 @@ def get_block(self, l_block):
-            raise LookupError(rffi.charp2str(llwrap.LLVMPrintValueToString(l_block)))
+            raise LookupError(self._get_block_string(l_block))
+            pass
+    @jit.elidable_promote()
```
Missing blank line.

ltratt commented on llvm_objects.py:
```diff
@@ -162,7 +205,7 @@ def _get_instructions(self, l_block, w_module):
             w_instrs.append(current_instr)
             l_instr = llwrap.LLVMGetNextInstruction(l_instr)
-        return w_instrs
+        return w_instrs[:]
```
Does everyone who call this function expect to mutate the list? If not, should the caller be responsible for cloning the list?

gabi-250 replied:
The list should never mutate. I'm returning a copy of the list because the return value of this method gets assigned to an immutable attribute. If I had returned the list itself, then the attribute wouldn't have been immutable.

vext01 replied:
You should be able to put the copy in either position, caller or callee, no? Or RPython chokes?

gabi-250 replied:
Hmm, I think you are right. I could change it back it if you feel that's a more sensible solution.

vext01 replied:
It's up to you.  Have a think...

ltratt replied:
Since `w_instrs` is a fresh list in this function (line 155), I don't see how "it's assigned to an immutable thingy" can be a problem (but maybe I'm missing out on something).

gabi-250 replied:
By doing `self.w_instructions = self._get_instructions(l_block, w_module)` you are assigning to `w_instructions` (which is an immutable field) the list you are creating in `_get_instructions`. The list created in `_get_instructions` is `w_instr`, which gets mutated. Therefore, RPython complains that you are mutating an immutable field (even though the list only gets changed before being assigned to an immutable field).

If I hadn't returned a copy of the list, I would've ended up with this error:
`[translation:ERROR] ListChangeUnallowed:`
`resizing list`
`...`

gabi-250 replied:
Maybe it's clearer if I put it this way:
A is an immutable field
B is a list I'm creating by appending values (it gets changed)
`A = B` would lead to an error: A and B point to the same object which is why RPython thinks you are mutating list A.
Please correct me if I am wrong or if I misunderstood your question.

ltratt replied:
Aha, I see. I agree that you have to do the copying, but I think it should be done by the person calling `_get_instructions` (since there's no particular reason why it should return an immutable list; it just depends on what the caller wants to do).

gabi-250 replied:
Well, even if it returned a copy, it wouldn't necessarily be an immutable list. I'm not sure why one way would be better than another. I would argue returning a copy is better, because the original version cannot be altered by the caller.

ltratt replied:
You really don't want functions making copies unless you know callers _always_ need copies -- it's a great way of introducing hard-to-spot inefficiencies.

gabi-250 replied:
Ah, I see. I'll make the caller copy the list.

vext01 commented:
Can you show us the benchmark times before and after?

vext01 commented on interpreter.py:
```diff
@@ -16,8 +19,12 @@ def jitpolicy(driver):
     from rpython.jit.codewriter.policy import JitPolicy
     return JitPolicy()
 
+def get_location(interpreter, block, function):
```
I'm nitpicking now, but why not call it `get_printable_location` which is standard RPython terminology.

gabi-250 replied:
I can always rename it.

gabi-250 commented:
Before any optimizations it took 0.058s to run both the iterative and the recursive fibonacci benchmark. Now it takes 0.062s for the recursive fibonacci benchmark to run and 0.056s for the iterative one. I should mention I'm running the tests 5 times (with multitime) and that the recursive algorithm always computes the 10th fibonacci number (technically the 11th, since the first fibonacci number according to my algorithm is 1, not 0) and the iterative one computes the 45th fibonacci number.

vext01 commented:
Hrm, that's quite an underwhelming optimisation.

Can we compute a larger fib number (one that causes more than a second runtime) in-case the short runtime is distorting the result? Or you could put a large loop around the existing benchmark.

gabi-250 commented:
I'll do that and let you know how it goes.

gabi-250 commented:
No optimizations:
fibonacci_rec: 2.054
fibonacci_iter: 0.64
After optimizations:
fibonacci_rec: 1.662
fibonacci_iter: 0.62

The trouble with the iterative benchmark is something I've noticed just now, unfortunately: the bytecode file contains the loop needed to calculate the fibonacci number, but it never gets executed, since the result is hard-coded in the `printf` statement. Therefore, I need to rewrite the benchmark so that clang doesn't make such optimizations.

gabi-250 committed:
Remove _attrs_, unnecessary method, some fields.

ltratt commented:
My comments are addressed, so we just need to see if @vext01 and @cfbolz are happy.

ltratt commented on llvm_objects.py:
```diff
@@ -244,31 +280,48 @@ class W_PhiInstruction(W_BaseInstruction):
+        self.incoming_block = None
+        self.incoming_value = self._get_incoming_value(l_instr)
+
+    def _get_incoming_value(self, l_instr):
```
Can this only ever be called from `__init__`? If so, I suggest merging it into the `__init__` method, as it's clearly quite tightly coupled (e.g. the `[:]` immutable copy thing).

gabi-250 replied:
Okay, I'll move the code inside `__init__`.

ltratt commented:
I think the use of `_attrs_` is pointless (except, admittedly, when executing in CPython), and should probably be removed.

gabi-250 commented:
I'm using `_attrs_` because there seems to be some kind of problem with inheritance. The base class ends up having all the attributes of its subclasses if I don't explicitly specify the attributes of each class (if I remember this correctly).

ltratt commented:
Can you try removing them and see what happens? Unless there's some weird magic going on, I don't know how the base class can possibly end up with the attributes of the subclass. We need to understand this a little better, at the very least.

ltratt commented on llvm_objects.py:
```diff
@@ -196,46 +242,36 @@ def _get_operands(self, l_instr, w_module):
     def __init__(self, l_instr, w_module):
         W_BaseInstruction.__init__(self, l_instr, w_module)
-        self.l_bb_uncond = llwrap.LLVMValueAsBasicBlock(self.l_operands[0])
+        self.w_bb_uncond = None
```
If `w_bb_uncond` is immutable, what's the point of the field?

gabi-250 replied:
`w_bb_uncond` gets initialized in `_set_instr_blocks`.

ltratt replied:
Although this _might_ be safe, it's generally dangerous to make a field immutable if it's initial value isn't its final one. [In practise, this is OK provided that the final value is written before it is read. But it's easy to get this wrong, and then the program behaves incorrectly.]

gabi-250 replied:
I see. So for the time being, should I leave it as it is?

ltratt replied:
I don't know if it's safe in this case. If you have verified that it is _definitely_ safe, and also know that it will always be safe in the future, keep it. Otherwise, unless you're 100% sure, remove the immutable field, as it's potentially very dangerous.

gabi-250 replied:
Well, `w_bb_uncond` won't be an immutable field anymore as I'm not entirely sure it's safe.

gabi-250 commented:
I've removed the `_attrs_` from the instruction classes, but some sort of weird magic did happen earlier, which is why I added `_attrs_` in the first place. @cfbolz mentioned last week: 

> no,  to prevent that you add _attrs_
> to make the attribute not move to the base class

This was comment was made when he explained that subclass attributes tend to get transferred to the base class. I'll remove them from everywhere and see if it runs correctly. I might have misunderstood this though.

ltratt commented:
Do things work OK now that the `_attrs_` are removed?

gabi-250 commented:
It seems to work without `_attrs_`, but this wasn't the case last week. It probably runs without `_attrs_` because I've made some other changes as well. Do you think it would be better if I were to leave `_attrs_` out entirely?

ltratt commented:
If `_attrs_` aren't needed now, let's remove them.

ltratt commented:
OK, let's squash as appropriate, then I'll merge.

gabi-250 commented:
Done.

ltratt commented:
The "Make caller copy the list." commit has a misleading title that needs changing.

gabi-250 commented:
I'll change it.

gabi-250 commented:
Hopefully it's better now.

ltratt commented:
Thanks!
